src/Eshopworld.WorkerProcess/LeaseAllocator.cs

```diff
@@ -34,83 +35,103 @@ public class LeaseAllocator : ILeaseAllocator
-
-                    await Task.Delay(delay).ConfigureAwait(false);
-
-                    // activate
-                    var now = ServerDateTime.UtcNow;
-                    lease.Interval = _slottedInterval.Calculate(ServerDateTime.UtcNow, _options.Value.LeaseInterval);
-                    lease.LeasedUntil = now.Add(lease.Interval.Value);
-
-                    updateResult = await _leaseStore.TryUpdateLeaseAsync(lease).ConfigureAwait(false);
+                    InstanceId = instanceId,
+                    Interval = interval,
+                    LeasedUntil = createTime.Add(interval),
+                    Priority = _options.Value.Priority,
+                    LeaseType = _options.Value.WorkerType
+                };
+                var createResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryCreateLeaseAsync).ConfigureAwait(false);
+                if (createResult.Result || createResult.Lease==null)
+                {
+                    return createResult.Lease;
+                }
 
-                    if (updateResult.Result)
-                    {
-                        _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                            _options.Value.Priority,
-                            $"Lease acquired. {GenerateLeaseInfo(updateResult.Lease)}"));
+                lease = createResult.Lease;
+            }
 
-                        return updateResult.Lease;
-                    }
 
-                    _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
-                        _options.Value.Priority,
-                        $"Lease activation failed lease already acquired. Allocated Lease Id: [{lease.Id}]"));
-                }
+            lease.Priority = _options.Value.Priority;
+            lease.InstanceId = instanceId;
+            var updateTime = ServerDateTime.UtcNow;
+            lease.Interval = _slottedInterval.Calculate(updateTime, _options.Value.LeaseInterval);
+            lease.LeasedUntil = updateTime.Add(lease.Interval.Value);
+            var updateResult = await TryAcquireLease(lease, instanceId, _leaseStore.TryUpdateLeaseAsync).ConfigureAwait(false);
+            if (updateResult.Result || updateResult.Lease == null)
+            {
+                return updateResult.Lease;
+            }
+            _telemetry.Publish(new LeaseAcquisitionEvent(instanceId, _options.Value.WorkerType,
+                _options.Value.Priority,
+                $"Lease failed to be acquired on update. {GenerateLeaseInfo(updateResult.Lease)}"));
+            return null;
+           
+        }
 
-                // Read the lease for event
-                lease = await _leaseStore.ReadByLeaseTypeAsync(_options.Value.WorkerType).ConfigureAwait(false);
+        private async Task<LeaseStoreResult> TryAcquireLease(ILease lease, Guid instanceId, Func<ILease,Task<LeaseStoreResult>> leaseAcquirer)
```

Comment:
Change lease store contract and make sure we can persist the lease always.
The lease store should either create new document or update existing (not sure if we have to match the etag as only winner will be updating the lease which will be the same for each instance of the winner - so does not matter who makes the update really).

Ideally, only one winner should update the lease (so etag could be used) but you will deal with that case here then - if update did not happen due to etag, you simply accept it and still return the winning lease (not null)

Reply:
do you mean have a method CreateOrUpdateLease in the LeaseStore?

Reply:
As both options discussed above are valid, this comment can be closed.